nixos/modules/services/networking/unbound.nix

```diff
@@ -6,41 +6,113 @@ let
 
   cfg = config.services.unbound;
 
-  stateDir = "/var/lib/unbound";
-
-  access = concatMapStringsSep "\n  " (x: "access-control: ${x} allow") cfg.allowedAccess;
-
-  interfaces = concatMapStringsSep "\n  " (x: "interface: ${x}") cfg.interfaces;
-
-  isLocalAddress = x: substring 0 3 x == "::1" || substring 0 9 x == "127.0.0.1";
-
-  forward =
-    optionalString (any isLocalAddress cfg.forwardAddresses) ''
-      do-not-query-localhost: no
-    '' +
-    optionalString (cfg.forwardAddresses != []) ''
-      forward-zone:
-        name: .
-    '' +
-    concatMapStringsSep "\n" (x: "    forward-addr: ${x}") cfg.forwardAddresses;
-
-  rootTrustAnchorFile = "${stateDir}/root.key";
-
-  trustAnchor = optionalString cfg.enableRootTrustAnchor
-    "auto-trust-anchor-file: ${rootTrustAnchorFile}";
-
-  confFile = pkgs.writeText "unbound.conf" ''
-    server:
-      directory: "${stateDir}"
-      username: unbound
-      chroot: "${stateDir}"
-      pidfile: ""
-      ${interfaces}
-      ${access}
-      ${trustAnchor}
-    ${cfg.extraConfig}
-    ${forward}
-  '';
+  settingsSubmodule = let
+    validConfigTypes = with types; either int (either str (either bool float));
+  in {
+    options = {
+      server = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = {
```

Comment:
You probably don't want to provide a `default` here because if the user sets any value in `server` the entire `default` will be overwritten... if I recall correctly. Instead you can add the following to the `config` section:
```
config = mkIf cfg.enable {
  services.unbound.server = {
    directory = cfg.stateDir;
    username = cfg.user;
    chroot = cfg.stateDir;
    pidfile = ''""'';
    interface = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
    access-control = [ "127.0.0.0/24 allow" ];
    auto-trust-anchor-file = if cfg.enableRootTrustAnchcor then rootTrustAnchorFile else null;
  };
}
```

Same applies for other options like `settings`, etc...

Reply:
Why not `directory = mkDefault cfg.stateDir` btw?

Reply:
I prefer `directory = cfg.stateDir` because if someone sets an option they should expect that to be the value, and other logic in the module might depend on that. The user always has the fallback of using `lib.mkForce` like so: `services.unbound.settings.directory = lib.mkForce "/something/else";`. That being said... maybe `pidfile` and `access-control` should use `mkDefault`. I think a judgement call by someone familiar with `unbound` is required.

Reply:
Fair enough :thumbsup:

Reply:
I would set `interface` and `access-control` with `mkDefault` to replicate previous behaviour. `pidfile` should be empty unless the user **really** wants to override that behaviour in which case they should use `mkForce`

Reply:
@rissson sounds reasonable :+1:

Reply:
After some more reflection, I also set `auto-trust-anchor-file` with `mkDefault`.